devlib/instrument/schedstats.py

```diff
@@ -0,0 +1,224 @@
+       of migrations successfully triggered by active_load_balance will be
+       colelcted for each sched domain, with a channel for each domain.
+
+    The measurements are named according to corresponding identifiers in the
+    kernel scheduler code. The names for ``sched_domain.lb_*`` stats, which are
+    recorded per ``cpu_idle_type`` are suffixed with a ':' followed by the idle
+    type, for example ``'lb_balanced:CPU_NEWLY_IDLE'``.
+
+    Only supports schedstats version 15.
+
+    Only supports the CPU and domain data in /proc/schedstat, not the per-task
+    data under /proc/<pid>/schedstat.
+    """
+
+    mode = INSTANTANEOUS
+
+    sysctl_path = '/proc/sys/kernel/sched_schedstats'
+    schedstat_path = '/proc/schedstat'
+
+    def __init__(self, *args, **kwargs):
+        super(SchedstatsInstrument, self).__init__(*args, **kwargs)
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+        # Check schedstats is present in the kernel and the format version
+        # matches what we can parse.
+        try:
+            lines = self.target.read_value(self.schedstat_path).splitlines()
+        except TargetError:
+            if not self.target.file_exists(self.schedstat_path):
+                raise TargetError('schedstats not supported by target. '
+                                  'Ensure CONFIG_SCHEDSTATS is enabled.')
+            raise
+
+        match = re.search(r'version ([0-9]+)', lines[0])
+        if not match or match.group(1) != '15':
+            raise TargetError(
+                'Unsupported schedstat version string: "{}"'.format(lines[0]))
+
+        # On 4.6+ kernels, schedstats needs to be enabled via kernel cmdline or
+        # sysctl. If not, we'll just get a load of zeroes.
+        self.old_sysctl_value = None
+        if self.target.kernel_version.parts >= (4, 6):
+            if self.target.file_exists(self.sysctl_path):
+                self.old_sysctl_value = self.target.read_int(self.sysctl_path)
+                self.target.write_value(self.sysctl_path, 1)
+            else:
+                try:
+                    cmdline = self.target.read_value('/proc/cmdline')
+                except TargetError:
+                    raise TargetError(
+                        "Couldn't verify that schedstats is enabled. "
+                        "Enabling CONFIG_PROC_SYSCTL will probably help")
+                if "schedstats=enable" not in cmdline:
+                    raise TargetError(
+                        "schedstats is compiled into the kernel but not enabled at runtime. "
+                        "Enable CONFIG_PROC_SYSCTL or add schedstats=enable to the cmdline.")
+
+        # Take a sample of the schedstat file to figure out which channels to
+        # create.
+        # We'll create a site for each CPU and a site for each sched_domain.
+        for site, measures in self._get_sample().iteritems():
+            if site.startswith('cpu'):
```

Comment:
Is it not more generic to parse the first column (up to excluding the ID) and create a category with that name?
Something like that should do the job:
```python
site_re = regexp.compile("^(?P<category>[a-z]+)(?P<id>[0-9]+)")
```

Reply:
IIRC I thought about that, but the thing is that `domain`s are a sub-category of `cpu`s, and there's no way to detect that: the parser has to be aware of this relationship. I can't imagine any other types of category (tasks? They have a separate schedstats file though. Task groups & sched entities would too, if they had schedstats), so I can't predict whether it makes sense to assume that the other category is a sub-entity of `cpu`s, a sub-category of `domain`s, or a new category entirely. Does that make sense?

Reply:
Ah OK now I understand you!

To be honest while it would be more generic I don't think this is worth it unless there are more categories.. even with the EAS stats there are still only 2 category types, and even with the addition of tasks there are only 3. I don't think it's worth adding a regex (which nobody likes reading) just to deal with that..